### PR TITLE
Port JS API to new Context API

### PR DESCRIPTION
**Pull request by epilys**
**All done.**

~~Status: some bugs in Memory tests.~~

~~Also js-wasi tests are failing.~~

#2892

### CODE REVIEW

File: lib/api/src/js/exports.rs [hunk 1]
```diff
@@ -141,51 +149,61 @@ impl Exports {
...
+    ) -> Result<T, ExportError>
     where
         Args: WasmTypeList,
         Rets: WasmTypeList,
         T: ExportableWithGenerics<'a, Args, Rets>,
     {
-        let out: T = self.get_with_generics(name)?;
+        let out: T = self.get_with_generics(ctx, name)?;
         Ok(out)
     }
```
Comment (syrusakbary) [inline lib/api/src/js/exports.rs:208]: I believe this doesn't exist any longer

Comment (epilys) [reply lib/api/src/js/exports.rs:208]: Ok, removing...

Comment (epilys) [reply lib/api/src/js/exports.rs:208]: Looks like it exists on the `sys` side, so keeping it.

File: lib/api/src/js/externals/function.rs [hunk 2]
```diff
@@ -302,65 +190,16 @@ impl Function {
...
-        let function = inner::Function::<Args, Rets>::new(func);
-        let address = function.address() as usize as u32;
-
-        let ft = wasm_bindgen::function_table();
-        let as_table = ft.unchecked_ref::<js_sys::WebAssembly::Table>();
-        let func = as_table.get(address).unwrap();
+        let binded_func = func.bind1(
+            &JsValue::UNDEFINED,
+            &JsValue::from_f64(ctx.as_raw() as *mut u8 as usize as f64),
+        );
```
Comment (syrusakbary) [inline lib/api/src/js/externals/function.rs:197]: Fix 1 (native functions)

File: lib/api/src/js/externals/function.rs [hunk 3]
```diff
@@ -1072,90 +955,46 @@ mod inner {
...
                         $( $x: FromToNativeWasmType, )*
                         Rets: WasmTypeList,
                         RetsAsResult: IntoResult<Rets>,
-                        Env: Sized,
-                        Func: Fn(&Env, $( $x ),* ) -> RetsAsResult + 'static
+                        Func: Fn(ContextMut<'_, T>, $( $x , )*) -> RetsAsResult + 'static,
                     {
-                        let env: &Env = unsafe { &*(ptr as *const u8 as *const Env) };
-                        let func: &Func = unsafe { &*(&() as *const () as *const Func) };
+                        let func: &Func = &*(&() as *const () as *const Func);
```
Comment (syrusakbary) [inline lib/api/src/js/externals/function.rs:981]: Fix 2 (native functions)

File: lib/api/tests/js_instance.rs [hunk 4]
```diff
@@ -437,14 +435,14 @@ mod js {
...
 
-        #[derive(Clone)]
+        #[derive(Clone, Debug)]
         struct Env {
             multiplier: u32,
             memory: Option<Memory>,
         }
 
         fn imported_fn(ctx: ContextMut<'_, Env>, arg: u32) -> u32 {
-            let memory: &Memory = panic!(); // FIXME?? ctx.memory_ref().unwrap();
```
Comment (syrusakbary) [inline lib/api/tests/js_instance.rs]: Fix (remove panic)

File: lib/api/tests/js_instance.rs [hunk 5]
```diff
@@ -550,14 +546,14 @@ mod js {
...
 
-        #[derive(Clone)]
+        #[derive(Clone, Debug)]
         struct Env {
             multiplier: u32,
             memory: Option<Memory>,
         }
 
         fn imported_fn(ctx: ContextMut<'_, Env>, args: &[Val]) -> Result<Vec<Val>, RuntimeError> {
-            let memory: &Memory = panic!(); //env.memory_ref().unwrap();
```
Comment (syrusakbary) [inline lib/api/tests/js_instance.rs]: Fix (remove panic)

**Pull request merged**